internals/secrethub/tree.go

```diff
@@ -47,42 +50,69 @@ func (cmd *TreeCommand) Register(r command.Registerer) {
 	clause := r.Command("tree", "List contents of a directory in a tree-like format.")
 	clause.Arg("dir-path", "The path to to show contents for").Required().PlaceHolder(optionalDirPathPlaceHolder).SetValue(&cmd.path)
 
+	clause.Flag("full-paths", "Print the full paths of the directories and secrets.").Short('f').BoolVar(&cmd.fullPaths)
+	clause.Flag("no-indentation", "Print the content without indentation.").Short('i').BoolVar(&cmd.noIndentation)
+	clause.Flag("no-report", "Skip the report at the bottom.").BoolVar(&cmd.noReport)
+
 	command.BindAction(clause, cmd.Run)
 }
 
 // printTree recursively prints the tree's contents in a tree-like structure.
-func printTree(t *api.Tree, w io.Writer) {
+func (cmd *TreeCommand) printTree(t *api.Tree, w io.Writer) {
 	name := colorizeByStatus(t.RootDir.Status, t.RootDir.Name)
-	fmt.Fprintf(w, "%s/\n", name)
-
-	printDirContentsRecursively(t.RootDir, "", w)
-
-	fmt.Fprintf(w,
-		"\n%s, %s\n",
-		pluralize("directory", "directories", t.DirCount()),
-		pluralize("secret", "secrets", t.SecretCount()),
-	)
+	fmt.Fprintf(w, "%s\n", name)
+
+	var format [4]string
+	if cmd.noIndentation {
+		format[0] = "%s%s%s\n"
+		format[1] = "%s%s%s\n"
+		format[2] = ""
+		format[3] = ""
+	} else {
+		format[0] = "%s└── %s%s\n"
+		format[1] = "%s├── %s%s\n"
+		format[2] = "    "
+		format[3] = "│   "
+	}
+	if cmd.fullPaths {
+		cmd.printDirContentsRecursively(t.RootDir, "", w, format, t.RootDir.Name)
+	} else {
+		cmd.printDirContentsRecursively(t.RootDir, "", w, format, "")
+	}
+	if !cmd.noReport {
+		fmt.Fprintf(w,
+			"\n%s, %s\n",
+			pluralize("directory", "directories", t.DirCount()),
+			pluralize("secret", "secrets", t.SecretCount()),
+		)
+	}
 }
 
 // printDirContentsRecursively is a recursive function that prints the directory's contents
 // in a tree-like structure, subdirs first followed by secrets.
-func printDirContentsRecursively(dir *api.Dir, prefix string, w io.Writer) {
+func (cmd *TreeCommand) printDirContentsRecursively(dir *api.Dir, prefix string, w io.Writer, format [4]string, prevPath string) {
 
 	sort.Sort(api.SortDirByName(dir.SubDirs))
 	sort.Sort(api.SortSecretByName(dir.Secrets))
 
 	total := len(dir.SubDirs) + len(dir.Secrets)
 
+	if !cmd.fullPaths {
+		prevPath = ""
+	} else {
+		prevPath += "/"
+	}
+
 	i := 0
 	for _, sub := range dir.SubDirs {
 		name := colorizeByStatus(sub.Status, sub.Name)
 
 		if i == total-1 {
-			fmt.Fprintf(w, "%s└── %s/\n", prefix, name)
-			printDirContentsRecursively(sub, prefix+"    ", w)
+			fmt.Fprintf(w, format[0], prefix, prevPath, name)
+			cmd.printDirContentsRecursively(sub, prefix+format[2], w, format, prevPath+name.(string))
```

Comment:
Because `name` is already colorized (made red) in case the directory is flagged, the directory part of the path of other secrets is also flagged.

What's more, in case secrets/directories are flagged, only part of the full path is printed in red.

I'll give an example to illustrate:

In this example with two secrets, only `simon/my-app/mailchimp/api-key` is flagged, `simon/my-app/stripe/api-key` is still okay. However, if I run `secrethub tree -f simon/test`, everything marked with astrixes becomes red:

    simon/**my-app**
    ├── simon/**my-app**/stripe
    │   └── simon/**my-app**/stripe/api-key
    └── simon/**my-app**/**mailchimp**
        └── simon/**my-app**/**mailchimp**/**api-key**

This is what I would expect instead:
```
    **simon/my-app**
    ├── simon/my-app/stripe
    │   └── simon/my-app/stripe/api-key
    └── **simon/my-app/mailchimp**
        └── **simon/my-app/mailchimp/api-key**
```

I believe we can achieve this by making two changes:
1. Evaluate `colorizeByStatus` on the full path instead of just the name
1. Evaluate `colarizeByStatus` only on the string passed to `fmt.Fprintf`, not on the string passed to the recursive call (`cmd.printDirContentsRecursively`)

Reply:
This should now be fixed.

Reply:
With prefixes (sometimes) colorized, this doesn't look very good yet. Please try and run this for yourselves to see the result.

You can create a flagged secret by [creating a service account](https://secrethub.io/docs/reference/cli/service/#init), reading a secret with the service account and then [revoking the service account](https://secrethub.io/docs/reference/cli/service/#revoke).

If you have any questions, let me know :smile:

Reply:
Thank you, @SimonBarendse, we now tested it on our machines and, to our knowledge, it works smoothly. Have a look and let us know if you can find any more issues.

Reply:
> we now tested it on our machines and, to our knowledge, it works smoothly

Just driving by here... this looks like something that should be possible to add an automated test for.

Reply:
Thanks @hculea, this indeed appears to be fixed now.

I like Marc's suggestion of adding a unit test for this. Could you do this?

Reply:
Of course!